src/ralph/models/selector.py

```diff
@@ -0,0 +1,128 @@
+
+    Args:
+        **filters: Each keyword represents the path to a field, the corresponding
+            value represents the value which the field should match.
+    """
+
+    return [Rule(LazyModelField(field), value) for field, value in filters.items()]
+
+
+class ModelSelector:
+    """Selects the matching model for a given event.
+
+    Attributes:
+        model_rules (dict): Stores the list of rules for each model.
+        decision_tree (dict): Stores the rule checking order for model selection.
+    """
+
+    model_rules: dict[Type[BaseModel], list[Rule]] = {}
+
+    def __init__(self, module="ralph.models.edx"):
+        """Instantiates ModelSelector."""
+
+        for object_ in import_module(module).__dict__.values():
+            if (
+                isinstance(object_, type)
+                and issubclass(object_, BaseModel)
+                and object_.__selector__
+            ):
+                self.model_rules[object_] = object_.__selector__
+        self.decision_tree = self.get_decision_tree(self.model_rules)
+
+    def get_model(self, event: dict, position=None):
+        """Recursively walks through the decision tree to return the matching model for the event.
+
+        Args:
+            event (dict): Event to retrieve the corresponding model.
+            position (dict): The position in the decision tree, `None` stands for "at the start".
+
+        Returns:
+            model (BaseModel): When event matches all rules of the model.
+            None: Otherwise.
+        """
+
+        if position is None:
+            position = self.decision_tree
+        rule = next(iter(position))
+        direction = rule.check(event)
+        next_position = position[rule][direction]
+        if not isinstance(next_position, dict):
+            return next_position
+        return self.get_model(event, next_position)
+
+    def get_decision_tree(self, model_rules):
+        """Recursively constructs the decision tree."""
+
+        rule_counter = Counter(chain.from_iterable(model_rules.values()))
+        if not rule_counter:
+            return {}
+
+        root_rule = rule_counter.most_common(1)[0][0]
+        true_model_rules = {}
+        false_model_rules = {}
+        for model, rules in model_rules.items():
+            if root_rule not in rules:
+                false_model_rules[model] = rules
+                continue
+            true_model_rules[model] = list(
+                filter(lambda rule: rule != root_rule, rules)
+            )
+
+        if next(iter(true_model_rules.values())) == []:
+            true_model_rules = next(iter(true_model_rules))
+        elif true_model_rules:
+            true_model_rules = self.get_decision_tree(true_model_rules)
+
+        if false_model_rules:
+            false_model_rules = self.get_decision_tree(false_model_rules)
+        else:
+            false_model_rules = None
+
+        return {root_rule: {True: true_model_rules, False: false_model_rules}}
```

Comment:
A general comment on the decision-tree like implementation: I think some variable names are not relevant and I find it hard to understand those pieces of code. I propose the following alternative algorithm:

1. Set/build the `decision_tree` during instantiation (and drop the `model_rules`) along with a rules registry (all `Rule` instances for a module, _e.g._ `edx`),
2. Given 3 models `A`, `B` and `C`, responding respectively to the following rules (`alpha`, `beta`), (`beta`), (`gamma`, `delta`), I propose the following decision tree structure:

```python
decision_tree = {
  alpha: {
    beta: A
  }, 
  beta: B, 
  gamma: {
    delta: C
  }, 
  delta: None
}
```

3. Recursively walking through this tree would stop when the node value is a model class, meaning that we've found the model corresponding to the input event.

WDYT?

Reply:
As discussed IRL, this proposal adds too much complexity in its implementation. We will stick to a binary decision tree.